Use PDM to manage macOS wheel dependencies

mwoehlke-kitware opened the pull request:
Use Drake's PDM-managed virtual environment (@python), rather than a separate virtual environment, to manage the additional Python bits that are needed to build a wheel. This allows us to pin the versions of these dependencies, for better reproducibility and is one less place where such things are being managed. Having the wheel dependencies installed "up front" is also in line with how Linux wheels are built.

Toward #22042.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/RobotLocomotion/drake/22437)
<!-- Reviewable:end -->

mwoehlke-kitware commented:
@drake-jenkins-bot mac-arm-sonoma-unprovisioned-clang-wheel-experimental-release please

mwoehlke-kitware commented:
+@jwnimmer-tri for feature review, please.





<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-wheel-experimental-release please

jwnimmer-tri reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> feature.  It took me a few false starts until I fully internalized which files were platform-specific and which files weren't, but once I understood that part it was all good.

> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-:-OGB_IAh33qYDirCbKjS:b-kva53k)* status: 2 unresolved discussions, needs at least two assigned reviewers

___
*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGB_mQ-2ZkxUOTton48:-OGB_mQ-2ZkxUOTton49:b-q5et9t) (no related file):*
BTW The `image/build-wheel.sh` file is linux-only but doesn't say that in a comment and has what appears to be a lot of dead "if linux" and "if darwin" branches.  That's probably worth cleaning up at some point soon.
___
*[`setup/python/pyproject.toml` line 22 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGBVnz059PccqYBww8N:-OGBVnz059PccqYBww8O:b-4tcot4) ([raw file](https://github.com/RobotLocomotion/drake/blob/5365b525f40c7a4dc58a4374b3265ed1cddbb4af/setup/python/pyproject.toml#L22)):*
> ```TOML
> ]
> 
> [dependency-groups]
> ```

nit This has grown enough complexity now that we need more maintainer commentary for this section explaining what's going on, covering:

(1) the "purpose" or "meaning" of each group, so people can decide when/whether to add something into it

(2) some at least minimal citations to the points of control that _select_ which group(s) are active, so developers can trace backwards to understand the effect of anything the edit here


<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
+@sammy-tri for platform review, please.





<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
@drake-jenkins-bot mac-arm-sonoma-unprovisioned-clang-wheel-experimental-release please

mwoehlke-kitware commented:
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-wheel-experimental-release please

jwnimmer-tri reviewed (commented):
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22437)* status: 1 unresolved discussion, LGTM missing from assignee sammy-tri(platform)

___
*[`setup/python/pyproject.toml` line 22 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGBVnz059PccqYBww8N:-OGWTp7b52rNJiMI1Ic5:b-u71j3h) ([raw file](https://github.com/RobotLocomotion/drake/blob/5365b525f40c7a4dc58a4374b3265ed1cddbb4af/setup/python/pyproject.toml#L22)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit This has grown enough complexity now that we need more maintainer commentary for this section explaining what's going on, covering:

(1) the "purpose" or "meaning" of each group, so people can decide when/whether to add something into it

(2) some at least minimal citations to the points of control that _select_ which group(s) are active, so developers can trace backwards to understand the effect of anything the edit here
</blockquote></details>

Remember to still fix this part.


<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
*[`setup/python/pyproject.toml` line 22 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGBVnz059PccqYBww8N:-OGWUJSl5S5-55E5xgDw:be9g0b3) ([raw file](https://github.com/RobotLocomotion/drake/blob/5365b525f40c7a4dc58a4374b3265ed1cddbb4af/setup/python/pyproject.toml#L22)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Remember to still fix this part.
</blockquote></details>

Hmm, this comment only just showed up.

I'd prefer to punt on it; this section is going to have at least two more PRs that could make these changes, but right now there isn't a way to do this without conflicting with #22378.


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22437)* status: 1 unresolved discussion, LGTM missing from assignee sammy-tri(platform)

___
*[`setup/python/pyproject.toml` line 22 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGBVnz059PccqYBww8N:-OGWVd_VAhe-mI7K-dBi:bz9z6rc) ([raw file](https://github.com/RobotLocomotion/drake/blob/5365b525f40c7a4dc58a4374b3265ed1cddbb4af/setup/python/pyproject.toml#L22)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

Hmm, this comment only just showed up.

I'd prefer to punt on it; this section is going to have at least two more PRs that could make these changes, but right now there isn't a way to do this without conflicting with #22378.
</blockquote></details>

Ideally either this PR or the other PR would add the commentary, but I suppose if there's a third PR in the wings to be opened in the next couple days, that's sufficient.


<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
*[`setup/python/pyproject.toml` line 22 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGBVnz059PccqYBww8N:-OGWes2-4zntWm9H4oXV:b-ogtryt) ([raw file](https://github.com/RobotLocomotion/drake/blob/5365b525f40c7a4dc58a4374b3265ed1cddbb4af/setup/python/pyproject.toml#L22)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Ideally either this PR or the other PR would add the commentary, but I suppose if there's a third PR in the wings to be opened in the next couple days, that's sufficient.
</blockquote></details>

Okay, with #22378 landed and this still waiting on platform, rebased and done.


<!-- Sent from Reviewable.io -->

sammy-tri reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  No issues from me, just waiting on CI.

> Reviewed 2 of 6 files at r1, 2 of 3 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-:-OGX2_lrBkvPwVquUqUU:bmu4jfo)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees sammy-tri(platform),jwnimmer-tri(platform)



<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22437)* status: 1 unresolved discussion

___
*[`setup/python/pyproject.toml` line 43 at r3](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGX69L51Me9efMHR73T:-OGX69L51Me9efMHR73U:b-gp5t1i) ([raw file](https://github.com/RobotLocomotion/drake/blob/83a7e65ef485b64d96c82d4587b57fbe23fb3ffe/setup/python/pyproject.toml#L43)):*
> ```TOML
>     "websockets",
> ]
> # Dependencies needed to build a Drake wheel.
> ```

BTW Are these in _addition_ to the `dependencies = [` listed above (i.e., does this add more deps, or completely reset to just these three).

If these are additive, then the comment is incomplete / misleading.

If these are the _only_ things installed into the venv, then I don't understand how the CMake build/install step during the wheel build works correctly (where does PyYAML come from in that case)?


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22437)* status: 1 unresolved discussion

___
*[`setup/python/pyproject.toml` line 43 at r3](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGX69L51Me9efMHR73T:-OGX6c6F16Nhwg8t-5RR:b-gw6mwq) ([raw file](https://github.com/RobotLocomotion/drake/blob/83a7e65ef485b64d96c82d4587b57fbe23fb3ffe/setup/python/pyproject.toml#L43)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Are these in _addition_ to the `dependencies = [` listed above (i.e., does this add more deps, or completely reset to just these three).

If these are additive, then the comment is incomplete / misleading.

If these are the _only_ things installed into the venv, then I don't understand how the CMake build/install step during the wheel build works correctly (where does PyYAML come from in that case)?
</blockquote></details>

I guess given that "test" seems to be additive with the base dependencies, then the only problem here is that the comments in `[dependency-groups]` fail to communicate that these are additive versus the baseline.

I also wonder if we _want_ the `pdm` group to be additive...


<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
*[`setup/python/pyproject.toml` line 43 at r3](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGX69L51Me9efMHR73T:-OGf8b4l1b9E4xTtk6lO:b-ouoc41) ([raw file](https://github.com/RobotLocomotion/drake/blob/83a7e65ef485b64d96c82d4587b57fbe23fb3ffe/setup/python/pyproject.toml#L43)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I guess given that "test" seems to be additive with the base dependencies, then the only problem here is that the comments in `[dependency-groups]` fail to communicate that these are additive versus the baseline.

I also wonder if we _want_ the `pdm` group to be additive...
</blockquote></details>

`\[dependency-groups\]`, when managed by PDM, are in addition unless explicitly excluded with `--no-default`; see [development only dependencies](https://pdm-project.org/en/latest/usage/dependency/#add-development-only-dependencies). I'll update the comments.

The `pdm` group is special; we never install it via `pdm sync`. It's there so we can `pdm export` that group from the lock file as a `requirements.txt`, and we *do* use `--no-default` when doing so.


<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
*[`setup/python/pyproject.toml` line 43 at r3](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGX69L51Me9efMHR73T:-OGfBwBaBLFHzEo6SMDZ:b-896fix) ([raw file](https://github.com/RobotLocomotion/drake/blob/83a7e65ef485b64d96c82d4587b57fbe23fb3ffe/setup/python/pyproject.toml#L43)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

`\[dependency-groups\]`, when managed by PDM, are in addition unless explicitly excluded with `--no-default`; see [development only dependencies](https://pdm-project.org/en/latest/usage/dependency/#add-development-only-dependencies). I'll update the comments.

The `pdm` group is special; we never install it via `pdm sync`. It's there so we can `pdm export` that group from the lock file as a `requirements.txt`, and we *do* use `--no-default` when doing so.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22437)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees sammy-tri(platform),jwnimmer-tri(platform)

___
*[`setup/python/pyproject.toml` line 43 at r3](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGX69L51Me9efMHR73T:-OGfCi0R4UfehH6oY0_9:brwj8v9) ([raw file](https://github.com/RobotLocomotion/drake/blob/83a7e65ef485b64d96c82d4587b57fbe23fb3ffe/setup/python/pyproject.toml#L43)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Okay that seems sensible, so it was only a comment/clarity problem.  However, I don't see any push yet with the new comments.


<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
*[`setup/python/pyproject.toml` line 43 at r3](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGX69L51Me9efMHR73T:-OGfDVZY7AICJUzswtJH:bxuyicn) ([raw file](https://github.com/RobotLocomotion/drake/blob/83a7e65ef485b64d96c82d4587b57fbe23fb3ffe/setup/python/pyproject.toml#L43)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Okay that seems sensible, so it was only a comment/clarity problem.  However, I don't see any push yet with the new comments.
</blockquote></details>

Oops, accidentally pushed to the wrong branch. Should be better now.


<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
*[`setup/python/pyproject.toml` line 43 at r3](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGX69L51Me9efMHR73T:-OGfFtQ87kxveCBqOAAo:b-ovk9id) ([raw file](https://github.com/RobotLocomotion/drake/blob/83a7e65ef485b64d96c82d4587b57fbe23fb3ffe/setup/python/pyproject.toml#L43)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

> The `pdm` group is special; we never install it via `pdm sync`. It's there so we can `pdm export` that group from the lock file as a `requirements.txt`, and we _do_ use `--no-default` when doing so.

Don't just tell me in reviewable.  Copy-paste this (or a rephrasing with equal entropy) into the comments.  Otherwise it's too high of a bar for people to tease out on their own.
</blockquote></details>

"Note that these are used to regenerate setup/python/requirements.txt and are installed via pip."

I don't want to document how `venv_upgrade` works in `pyproject.toml`. I added a note to refer to the script, and added additional documentation to the script.


<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
@drake-jenkins-bot mac-arm-sonoma-unprovisioned-clang-wheel-experimental-release please
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-wheel-experimental-release please

(Not expecting surprises, but since `master` has changed in the mean time...)

jwnimmer-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22437)* status: 1 unresolved discussion

___
*[`setup/python/pyproject.toml` line 43 at r3](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGX69L51Me9efMHR73T:-OGfHAy2A7j47aRi6lSE:b-a8pgva) ([raw file](https://github.com/RobotLocomotion/drake/blob/83a7e65ef485b64d96c82d4587b57fbe23fb3ffe/setup/python/pyproject.toml#L43)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

"Note that these are used to regenerate setup/python/requirements.txt and are installed via pip."

I don't want to document how `venv_upgrade` works in `pyproject.toml`. I added a note to refer to the script, and added additional documentation to the script.
</blockquote></details>

Well, this commit isn't going to merge until the code is clear instead of confusing.  I appreciate the desire to "not repeat yourself" -- that is a good habit -- but sometimes in documentation we need be a little repetitive, to help people along. We don't need the repeat everything, but the key fact that the pdm group does _not_ incorporate the default list of `dependencies` must appear here, in some form.  That fact is an _invariant_ of this design, so will not bitrot in case the sync code changes.


<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
*[`setup/python/pyproject.toml` line 43 at r3](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGX69L51Me9efMHR73T:-OGfMUzF2IQ-gHgeSIAx:b-1wwwlo) ([raw file](https://github.com/RobotLocomotion/drake/blob/83a7e65ef485b64d96c82d4587b57fbe23fb3ffe/setup/python/pyproject.toml#L43)):*
> that the pdm group does _not_ incorporate the default list of `dependencies` must appear here

I don't understand this. Why would anyone expect "Dependencies needed to manage Drake's dependencies" to include "Dependencies needed to build Drake", especially when the other groups are explicitly labeled as "*Additional* dependencies ..."?

> That fact is an _invariant_ of this design, so will not bitrot in case the sync code changes.

I don't understand this.


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22437)* status: 1 unresolved discussion

___
*[`setup/python/pyproject.toml` line 43 at r3](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGX69L51Me9efMHR73T:-OGfbOWkDx9SoXeCSFOw:b-pdt6cn) ([raw file](https://github.com/RobotLocomotion/drake/blob/83a7e65ef485b64d96c82d4587b57fbe23fb3ffe/setup/python/pyproject.toml#L43)):*
> Why would anyone expect "Dependencies needed to manage Drake's dependencies" to include "Dependencies needed to build Drake", especially when the other groups are explicitly labeled as "_Additional_ dependencies ..."?

Because not everyone will read the tea leaves that "Additional" is so special that not saying "Additional" in one place is supposed to communicate they they are not additional in that case.  As you say PDM by default _does_ include the `dependencies` as a baseline for all the groups.  The fact that we carefully opt-out of that baseline behavior every single time we use this particular group is **not obvious** here.

Of course _some_ people could draw the conclusion without any extra comments, because it would be _odd_ to include e.g. matplotlib in our venv managment tools.  However, **it was not immediately clear to me** so we **need more comments**.  It will also be unclear to some future fraction of Drake's maintainers.  The cost of a comment is vanishingly small, and the benefit is potentiality large.

Here's a concrete suggestion.

> Dependencies needed to manage Drake's dependencies. Note that these are used to regenerate setup/python/requirements.txt (without including the baseline `dependencies` list above) and are installed via pip.

---

> > That fact is an _invariant_ of this design, so will not bitrot in case the sync code changes.
> 
> I don't understand this.

That's fine, it's tangential and we can ignore it.


<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
*[`setup/python/pyproject.toml` line 43 at r3](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGX69L51Me9efMHR73T:-OGflfrD50k85UHmvr4E:b7geka) ([raw file](https://github.com/RobotLocomotion/drake/blob/83a7e65ef485b64d96c82d4587b57fbe23fb3ffe/setup/python/pyproject.toml#L43)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

> Why would anyone expect "Dependencies needed to manage Drake's dependencies" to include "Dependencies needed to build Drake", especially when the other groups are explicitly labeled as "_Additional_ dependencies ..."?

Because not everyone will read the tea leaves that "Additional" is so special that not saying "Additional" in one place is supposed to communicate they they are not additional in that case.  As you say PDM by default _does_ include the `dependencies` as a baseline for all the groups.  The fact that we carefully opt-out of that baseline behavior every single time we use this particular group is **not obvious** here.

Of course _some_ people could draw the conclusion without any extra comments, because it would be _odd_ to include e.g. matplotlib in our venv managment tools.  However, **it was not immediately clear to me** so we **need more comments**.  It will also be unclear to some future fraction of Drake's maintainers.  The cost of a comment is vanishingly small, and the benefit is potentiality large.

Here's a concrete suggestion.

> Dependencies needed to manage Drake's dependencies. Note that these are used to regenerate setup/python/requirements.txt (without including the baseline `dependencies` list above) and are installed via pip.

---

> > That fact is an _invariant_ of this design, so will not bitrot in case the sync code changes.
> 
> I don't understand this.

That's fine, it's tangential and we can ignore it.
</blockquote></details>

Alternatively, let's just not conflate these at all. See #22473.


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22437)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees sammy-tri(platform),jwnimmer-tri(platform)

___
*[`setup/python/pyproject.toml` line 43 at r3](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGX69L51Me9efMHR73T:-OGfq-ff4Yzt0HkAuzpX:b-qddqw5) ([raw file](https://github.com/RobotLocomotion/drake/blob/83a7e65ef485b64d96c82d4587b57fbe23fb3ffe/setup/python/pyproject.toml#L43)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

Alternatively, let's just not conflate these at all. See #22473.
</blockquote></details>

OK since this will be solve by that PR, I'll close the thread and you can land this one first if you prefer.


<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
*[`setup/python/pyproject.toml` line 43 at r3](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-OGX69L51Me9efMHR73T:-OGfuW2cCi0bo37USkI9:b-3898zg) ([raw file](https://github.com/RobotLocomotion/drake/blob/83a7e65ef485b64d96c82d4587b57fbe23fb3ffe/setup/python/pyproject.toml#L43)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

OK since this will be solve by that PR, I'll close the thread and you can land this one first if you prefer.
</blockquote></details>

Let's land that one, since the resolution here depends on that, then this can remove the unnecessary documentation. (They're going to conflict in either order.)


<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
@drake-jenkins-bot mac-arm-sonoma-unprovisioned-clang-wheel-experimental-release please
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-wheel-experimental-release please

jwnimmer-tri commented:
FYI the final edits are small enough that we don't need to wait for Sam to re-look before merging this.

mwoehlke-kitware reviewed (commented):
Oh, agreed, I'm mainly waiting on CI. (Partly because I forgot to check that the previous run passed before pushing again... :flushed:)

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22437#-:-OGkKqen9MFr5bvOsTs5:btfprgn)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees sammy-tri(platform),jwnimmer-tri(platform)



<!-- Sent from Reviewable.io -->
